src/derivative_utils.jl

```diff
@@ -261,23 +263,29 @@ end
 @noinline _throwWJerror(W, J) = throw(DimensionMismatch("W: $(axes(W)), J: $(axes(J))"))
 @noinline _throwWMerror(W, mass_matrix) = throw(DimensionMismatch("W: $(axes(W)), mass matrix: $(axes(mass_matrix))"))
 
-@inline function jacobian2W!(W, mass_matrix, dtgamma, J, W_transform)::Nothing
+@inline function jacobian2W!(W::AbstractMatrix, mass_matrix::MT, dtgamma::Number, J::AbstractMatrix, W_transform::Bool)::Nothing where MT
   # check size and dimension
   iijj = axes(W)
   @boundscheck (iijj === axes(J) && length(iijj) === 2) || _throwWJerror(W, J)
   mass_matrix isa UniformScaling || @boundscheck axes(mass_matrix) === axes(W) || _throwWMerror(W, mass_matrix)
   @inbounds if W_transform
-    invdtgamma′ = inv(dtgamma)
-    for i in iijj[1]
-      @inbounds for j in iijj[2]
-        W[i, j] = muladd(mass_matrix[i, j], invdtgamma′, -J[i, j])
+    invdtgamma = inv(dtgamma)
+    if MT <: UniformScaling
+      copyto!(W, J)
+      @simd for i in diagind(W)
+          W[i] = muladd(-mass_matrix.λ, invdtgamma, J[i])
```

Comment:
what if the diagonal term is zero in a sparse matrix? That could happen

Reply:
This branch is only for dense matrices.

Reply:
how so?